File: cpp/tests/test_potential_fields.cpp
#include <cmath>
#include <sstream>
#include <string>

#include <gtest/gtest.h>

#include "navigation/core/params.hpp"
#include "navigation/core/trace.hpp"
#include "navigation/core/types.hpp"
#include "navigation/local_planning/reactive/potential_fields.hpp"
#include "navigation/local_planning/simulation.hpp"
#include "navigation/maps/loader.hpp"
#include "navigation/maps/occupancy_grid.hpp"
#include "test_util.hpp"

using namespace navigation;
using core::LocalTask;
using core::ParamSet;
using core::Pose;
using core::RobotState;
using core::TraceRecorder;
using local_planning::PotentialFieldsPlanner;
using local_planning::SimConfig;
using local_planning::SimStatus;
using local_planning::simulate;

namespace {

ParamSet real_config() {
  return ParamSet::from_yaml(test::repo_path("configs/local_planning/potential_fields.yaml"));
}

SimConfig sim_config_from(const ParamSet& params) {
  return SimConfig{params.get_float("control_dt"),     params.get_int("max_steps"),
                   params.get_float("goal_tolerance"), params.get_float("footprint_radius"),
                   params.get_int("stall_window"),     params.get_float("stall_distance")};
}

maps::OccupancyGrid2D& as_grid(core::MapBase& m) {
  return *dynamic_cast<maps::OccupancyGrid2D*>(&m);
}

local_planning::SimResult run_scenario(const std::string& map_name, const std::string& scenario_name) {
  auto params = real_config();
  auto map = maps::load_map(test::repo_path("maps/grid/" + map_name + ".yaml"), 0, 8);
  auto& grid = as_grid(*map);
  maps::Scenario sc = maps::load_scenario(test::repo_path("maps/scenarios/" + scenario_name + ".yaml"));
  PotentialFieldsPlanner planner(params);
  RobotState start{Pose{sc.start.x, sc.start.y, sc.start_theta}, 0.0, 0.0};
  LocalTask task{Pose{sc.goal.x, sc.goal.y, sc.goal_theta}, {}};
  return simulate(planner, grid, start, task, sim_config_from(params), nullptr);
}

// Locates `"key":<number>` in a raw trace line and parses the number -- the
// scalar counterpart of the array-field extraction helper in test_sampling.cpp.
double find_data_value(const std::string& line, const std::string& key) {
  std::string needle = "\"" + key + "\":";
  size_t at = line.find(needle);
  if (at == std::string::npos) return std::nan("");
  size_t start = at + needle.size();
  size_t end = line.find_first_of(",}", start);
  return std::stod(line.substr(start, end - start));
}

}  // namespace

// (a) reaches goal on open/clutter maps without collision -----------------------

TEST(PotentialFields, ReachesGoalOnOpenMapWithoutCollision) {
  auto r = run_scenario("open01", "open01_s1");
  EXPECT_EQ(r.status, SimStatus::REACHED);
  EXPECT_TRUE(r.success);
  EXPECT_GT(r.min_clearance, 0.0);
}

TEST(PotentialFields, ReachesGoalOnClutterMapWithoutCollision) {
  auto r = run_scenario("clutter01", "clutter01_s1");
  EXPECT_EQ(r.status, SimStatus::REACHED);
  EXPECT_TRUE(r.success);
  EXPECT_GT(r.min_clearance, 0.0);
}

// (b) U-trap: correct local-minimum stall, not a bug ----------------------------

TEST(PotentialFields, StallsInUTrapLocalMinimum) {
  // pf_trap01_s1: the goal sits straight through the U-trap's back wall, in
  // line with the opening. The attractive pull drives the robot in, and at
  // this row the top/bottom wall repulsion is symmetric so it cancels the
  // attractive pull head-on against the back wall -- a textbook local
  // minimum (Khatib 1986), not a simulator or tuning defect.
  auto r = run_scenario("pf_trap01", "pf_trap01_s1");
  EXPECT_EQ(r.status, SimStatus::STALLED);
  EXPECT_FALSE(r.success);
}

// (c) param validation failure ---------------------------------------------------

TEST(PotentialFields, BadKAttThrows) {
  std::string bad = test::write_temp(
      "potential_fields.yaml",
      "algorithm: potential_fields\ncategory: local_planning\nparams:\n"
      "  - name: k_att\n    type: float\n    default: 200.0\n    min: 0.001\n    max: 100.0\n"
      "    description: over max\n");
  EXPECT_THROW(ParamSet::from_yaml(bad), std::runtime_error);
}

// (behavior) repulsive force points away from a single obstacle -----------------

TEST(PotentialFields, RepulsiveForcePointsAwayFromObstacle) {
  // Single obstacle at ascii row 2 / col 2 -> world center (2.5, 2.5) (see
  // OccupancyGrid2D::cell_to_world). Robot sits one cell due north (2.5, 3.5),
  // so FIRAS repulsion must point further north (+y), away from the obstacle
  // -- an influence_radius just above the 1.0 m gap keeps this the only
  // source, isolating F_rep from the (goal-independent) attractive term.
  auto grid = test::make_grid({".......", ".......", "..#....", ".......", "......."});
  std::string cfg_path = test::write_temp(
      "potential_fields.yaml",
      "algorithm: potential_fields\ncategory: local_planning\nparams:\n"
      "  - {name: k_att, type: float, default: 0.5, min: 0.001, max: 100, description: a}\n"
      "  - {name: k_rep, type: float, default: 1.0, min: 0.0, max: 100, description: r}\n"
      "  - {name: influence_radius, type: float, default: 1.05, min: 0.05, max: 20, description: i}\n"
      "  - {name: k_v, type: float, default: 1.0, min: 0.001, max: 100, description: v}\n"
      "  - {name: k_omega, type: float, default: 4.0, min: 0.001, max: 100, description: o}\n"
      "  - {name: max_speed, type: float, default: 0.8, min: 0.01, max: 10, description: s}\n"
      "  - {name: max_omega, type: float, default: 5.0, min: 0.01, max: 20, description: w}\n"
      "  - {name: control_dt, type: float, default: 0.1, min: 0.001, max: 1, description: dt}\n"
      "  - {name: max_steps, type: int, default: 1000, min: 1, max: 100000, description: n}\n"
      "  - {name: goal_tolerance, type: float, default: 0.3, min: 0.01, max: 5, description: g}\n"
      "  - {name: footprint_radius, type: float, default: 0.2, min: 0.01, max: 5, description: f}\n"
      "  - {name: stall_window, type: int, default: 20, min: 1, max: 100000, description: sw}\n"
      "  - {name: stall_distance, type: float, default: 0.05, min: 0.0, max: 5, description: sd}\n");
  auto params = ParamSet::from_yaml(cfg_path);
  PotentialFieldsPlanner planner(params);
  RobotState state{Pose{2.5, 3.5, 0.0}, 0.0, 0.0};
  LocalTask task{Pose{2.5, 3.5, 0.0}, {}};  // coincident with start: F_att == 0

  std::ostringstream os;
  TraceRecorder rec(os);
  planner.compute_command(grid, state, task, params.get_float("control_dt"), &rec);

  std::istringstream in(os.str());
  std::string line;
  std::getline(in, line);
  ASSERT_NE(line.find("\"force_computed\""), std::string::npos);
  EXPECT_NEAR(find_data_value(line, "fx_rep"), 0.0, 1e-9);
  EXPECT_GT(find_data_value(line, "fy_rep"), 0.0);  // pushed north, away from the obstacle
}

File: python/tests/test_potential_fields.py
"""Potential Fields (Khatib 1986): goal-seek + obstacle-avoidance behavior via
the closed-loop simulator, including the well-known local-minimum stall."""

from __future__ import annotations

import io
import json
from pathlib import Path
from typing import Any

import pytest
import yaml
from conftest import REPO_ROOT, grid_from

from navigation.core.params import ParamError, ParamSet
from navigation.core.trace import TraceRecorder
from navigation.core.types import LocalTask, RobotState
from navigation.local_planning.reactive.potential_fields import PotentialFields
from navigation.local_planning.simulation import SimConfig, SimResult, SimStatus, simulate
from navigation.maps.loader import load_map, load_scenario

_ALGO = "potential_fields"
_CONFIG_PATH = REPO_ROOT / "configs" / "local_planning" / f"{_ALGO}.yaml"
_MAPS_DIR = REPO_ROOT / "maps"


def _config(tmp_path: Path, **overrides: Any) -> ParamSet:
    """Real config with selected defaults overridden."""
    doc = yaml.safe_load(_CONFIG_PATH.read_text())
    for entry in doc["params"]:
        if entry["name"] in overrides:
            entry["default"] = overrides[entry["name"]]
    out = tmp_path / f"{_ALGO}.yaml"
    out.write_text(yaml.safe_dump(doc), encoding="utf-8")
    return ParamSet.from_yaml(out)


def _sim_config(params: ParamSet) -> SimConfig:
    return SimConfig(
        control_dt=params.get_float("control_dt"),
        max_steps=params.get_int("max_steps"),
        goal_tolerance=params.get_float("goal_tolerance"),
        footprint_radius=params.get_float("footprint_radius"),
        stall_window=params.get_int("stall_window"),
        stall_distance=params.get_float("stall_distance"),
    )


def _run(tmp_path: Path, map_name: str, scenario_name: str, **overrides: Any) -> SimResult:
    params = _config(tmp_path, **overrides)
    grid = load_map(_MAPS_DIR / "grid" / f"{map_name}.yaml")
    scenario = load_scenario(_MAPS_DIR / "scenarios" / f"{scenario_name}.yaml")
    planner = PotentialFields(params)
    start = RobotState(pose=(scenario.start[0], scenario.start[1], scenario.start_theta))
    task = LocalTask(goal=(scenario.goal[0], scenario.goal[1], scenario.goal_theta))
    return simulate(planner, grid, start, task, _sim_config(params))


# --- (a) reaches goal on open/clutter maps, never collides --------------------
@pytest.mark.parametrize(
    ("map_name", "scenario_name"), [("open01", "open01_s1"), ("clutter01", "clutter01_s1")]
)
def test_reaches_goal_without_collision(
    tmp_path: Path, map_name: str, scenario_name: str
) -> None:
    result = _run(tmp_path, map_name, scenario_name)
    assert result.status is SimStatus.REACHED
    assert result.success is True
    assert result.min_clearance > 0.0


# --- (b) U-trap: correct local-minimum stall, not a bug ------------------------
def test_stalls_in_u_trap_local_minimum() -> None:
    # pf_trap01_s1: the goal sits straight through the U-trap's back wall, in
    # line with the opening. The attractive pull drives the robot in, and at
    # this row the top/bottom wall repulsion is symmetric so it cancels the
    # attractive pull head-on against the back wall -- a textbook local
    # minimum (Khatib 1986), not a simulator or tuning defect.
    grid = load_map(_MAPS_DIR / "grid" / "pf_trap01.yaml")
    scenario = load_scenario(_MAPS_DIR / "scenarios" / "pf_trap01_s1.yaml")
    params = ParamSet.from_yaml(_CONFIG_PATH)
    planner = PotentialFields(params)
    start = RobotState(pose=(scenario.start[0], scenario.start[1], scenario.start_theta))
    task = LocalTask(goal=(scenario.goal[0], scenario.goal[1], scenario.goal_theta))
    result = simulate(planner, grid, start, task, _sim_config(params))
    assert result.status is SimStatus.STALLED
    assert result.success is False


# --- (c) out-of-range param rejected at load time ------------------------------
def test_out_of_range_k_att_rejected(tmp_path: Path) -> None:
    doc = yaml.safe_load(_CONFIG_PATH.read_text())
    for entry in doc["params"]:
        if entry["name"] == "k_att":
            entry["default"] = entry["max"] + 1.0
    bad = tmp_path / f"{_ALGO}.yaml"
    bad.write_text(yaml.safe_dump(doc), encoding="utf-8")
    with pytest.raises(ParamError):
        ParamSet.from_yaml(bad)


# --- (behavior) repulsive force points away from a single obstacle ------------
def test_repulsive_force_points_away_from_obstacle(tmp_path: Path) -> None:
    # Single obstacle at ascii row 2 / col 2 -> world center (2.5, 2.5) (see
    # OccupancyGrid2D.cell_to_world). Robot sits one cell due north (2.5, 3.5),
    # so the FIRAS repulsion must point further north (+y), away from the
    # obstacle -- an influence_radius just above the 1.0 m gap keeps this the
    # only source, isolating F_rep from the (goal-independent) attractive term.
    grid = grid_from(
        [
            ".......",
            ".......",
            "..#....",
            ".......",
            ".......",
        ]
    )
    params = _config(tmp_path, influence_radius=1.05, k_rep=1.0)
    planner = PotentialFields(params)
    state = RobotState(pose=(2.5, 3.5, 0.0))
    task = LocalTask(goal=(2.5, 3.5, 0.0))  # coincident with start: F_att == 0

    buf = io.StringIO()
    recorder = TraceRecorder(buf)
    planner.compute_command(grid, state, task, params.get_float("control_dt"), recorder)

    events = [json.loads(line) for line in buf.getvalue().splitlines()]
    force_events = [e for e in events if e["event"] == "force_computed"]
    assert len(force_events) == 1
    data = force_events[0]["data"]
    assert data["fx_rep"] == pytest.approx(0.0, abs=1e-9)
    assert data["fy_rep"] > 0.0  # pushed north, away from the obstacle to the south
